tempodb/backend/gcs/gcs.go

```diff
@@ -186,6 +190,112 @@ func (rw *readerWriter) List(ctx context.Context, keypath backend.KeyPath) ([]st
 	return objects, nil
 }
 
+// ListBlocks implements backend.Reader
+func (rw *readerWriter) ListBlocks(ctx context.Context, keypath backend.KeyPath) (blockIDs []uuid.UUID, compactedBlockIDs []uuid.UUID, err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "readerWriter.ListBlocks")
+	defer span.Finish()
+
+	keypath = backend.KeyPathWithPrefix(keypath, rw.cfg.Prefix)
+	prefix := path.Join(keypath...)
+	if len(prefix) > 0 {
+		prefix += "/"
+	}
+
+	bb := blockboundary.CreateBlockBoundaries(rw.cfg.ListBlocksConcurrency)
+
+	errChan := make(chan error, len(bb))
+	wg := sync.WaitGroup{}
+	mtx := sync.Mutex{}
+
+	var min uuid.UUID
+	var max uuid.UUID
+
+	for i := 0; i < len(bb)-1; i++ {
+		min = uuid.UUID(bb[i])
+		max = uuid.UUID(bb[i+1])
+
+		wg.Add(1)
+		go func(min, max uuid.UUID) {
+			defer wg.Done()
+
+			iter := rw.bucket.Objects(ctx, &storage.Query{
+				Prefix:      prefix,
+				Delimiter:   "",
+				Versions:    false,
+				StartOffset: prefix + min.String(),
+			})
+
+			var parts []string
+			var id uuid.UUID
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					attrs, err := iter.Next()
+					if errors.Is(err, iterator.Done) {
+						return
+					}
+					if err != nil {
+						errChan <- fmt.Errorf("iterating blocks: %w", err)
+						return
+					}
+
+					parts = strings.Split(attrs.Name, "/")
+					// ie: <tenant>/<blockID>/meta.json
+					if len(parts) != 3 {
+						continue
+					}
+
+					if parts[2] != backend.MetaName && parts[2] != backend.CompactedMetaName {
+						continue
+					}
+
+					id, err = uuid.Parse(parts[1])
+					if err != nil {
+						continue
+					}
+
+					x := bytes.Compare(id[:], min[:])
+					if x < 0 {
+						return
+					}
+
+					y := bytes.Compare(id[:], max[:])
+					if y > 0 {
+						return
+					}
+
+					switch parts[2] {
+					case backend.MetaName:
+						mtx.Lock()
```

Comment:
generally very detailed locking like this should be avoided unless absolutely necessary. ideal is:
```
Lock()
defer Unlock()
```
but sometimes that's unwieldy in its own way. I think a nice compromise here is to lock before the switch and unlock after.

this comment applies to the s3 implementation as well.

Reply:
If we lock before the switch, then we lock just to check if there is match, which is on every object even if we won't end up making a modification that requires the lock.  I suppose we could check if either names are matched and then lock and then switch, but that seems less readable.  I'm find with making the change here, but I don't think its unreadable as is, since we lock only for one line, and then unlock.  I agree about the general pattern, but in this case I think it warranted the specific locks.  Given this, would you still prefer the lock + defer?

Reply:
Also defer in a for loop inside a go routine seems a little more dangerous to me.  So explicit lock + unlock seems like the right approach.

Reply:
> If we lock before the switch, then we lock just to check if there is match, which is on every object even if we won't end up making a modification that requires the lock.

You're checking a few lines up if this a meta or a compacted meta. So we should already know it's a match.

> Also defer in a for loop inside a go routine seems a little more dangerous to me. So explicit lock + unlock seems like the right approach.

That's why I said: "I think a nice compromise here is to lock before the switch and unlock after."

Reply:
Sorry you're right, you did say that.

Reply:
I've pushed a change for this.